### PR TITLE
Round #2 of LibSsl shim work

### PR DESCRIPTION
**Pull request by eerhardt**
Add shims for the next set of SSL interop methods.
I also moved the SSL SafeHandle classes out of the Interop class to be
better aligned with the rest of the SafeHandles.

Instead of having managed code duplicate the SSL_CIPHER struct, moving all the code that uses the cipher struct into a native shim method.

That also pulled a lot of algorithm constants to be defined in native instead of managed.

@bartonjs @stephentoub @vijaykota @rajansingh10 @shrutigarg @nguerrera

### CODE REVIEW

File: src/Common/src/Interop/Unix/libssl/Interop.SafeSslHandle.cs [hunk 1]
```diff
@@ -1,142 +0,0 @@
-// Copyright (c) Microsoft. All rights reserved.
```
Comment (vijaykota) [inline src/Common/src/Interop/Unix/libssl/Interop.SafeSslHandle.cs]: Nit: If file contents haven't changed, you might want to use "git mv"

File: src/Native/System.Security.Cryptography.Native/pal_ssl.cpp [hunk 2]
```diff
@@ -41,6 +43,256 @@ extern "C" SSL_CTX* SslCtxCreate(SSL_METHOD* method)
...
+{
+    return SSL_get_version(ssl);
+}
+
+static CipherAlgorithmType MapCipherAlgorithmType(const SSL_CIPHER* cipher)
+{
+    unsigned long enc;
+
+#ifdef HAVE_SSL_CIPHER_ALGORITHMS
+    const unsigned long SSL_ENC_MASK = 0x1C3F8000L;
```
Comment (vijaykota) [inline src/Native/System.Security.Cryptography.Native/pal_ssl.cpp]: Doubt: Doesn't this macro already exist in ssl header files?

File: src/Native/System.Security.Cryptography.Native/pal_ssl.cpp [hunk 3]
```diff
@@ -41,6 +43,256 @@ extern "C" SSL_CTX* SslCtxCreate(SSL_METHOD* method)
...
+            return HashAlgorithmType::Sha1;
+
+#ifndef HAVE_SSL_CIPHER_ALGORITHMS
+        case SSL_DataHashAlgorithm::SSL_GOST94:
+            return HashAlgorithmType::SSL_GOST94;
+
+        case SSL_DataHashAlgorithm::SSL_GOST89MAC:
+            return HashAlgorithmType::SSL_GOST89;
+
+        case SSL_DataHashAlgorithm::SSL_SHA256:
```
Comment (vijaykota) [inline src/Native/System.Security.Cryptography.Native/pal_ssl.cpp]: Since this is under ifndef, does it mean that OpenSsl 1.0.* doesn't support SHA256, SHA384 and AEAD?

File: src/Common/src/Interop/Unix/libssl/SslConnectionInfo.cs [hunk 4]
```diff
@@ -2,50 +2,34 @@
...
-        internal const int SSL_kGOST = 229391;
-        internal const int SSL_kSRP = 229392;
-        internal const int SSL_kKRB5 = 229393;
-
-        internal const int SSL_GOST94 = 229410;
-        internal const int SSL_GOST89 = 229411;
-        internal const int SSL_AEAD = 229412;
-
-        public readonly int Protocol;
+        public readonly SslProtocols Protocol;
```
Comment (vijaykota) [inline src/Common/src/Interop/Unix/libssl/SslConnectionInfo.cs]: Note that this will cause a subtle diff. in the Windows and Unix definitions for SslConnectionInfo

**Comment by vijaykota** [comment]
LGTM

File: src/Native/System.Security.Cryptography.Native/pal_err.cpp [hunk 5]
```diff
@@ -6,15 +6,28 @@
...
+}
+
+extern "C" uint64_t ErrGetErrorAlloc(int32_t* isAllocFailure)
 {
     unsigned long err = ERR_get_error();
 
-    *isAllocFailure = ERR_GET_REASON(err) == ERR_R_MALLOC_FAILURE;
+    if (isAllocFailure)
+    {
+        *isAllocFailure = ERR_GET_REASON(err) == ERR_R_MALLOC_FAILURE;
```
Comment (stephentoub) [inline src/Native/System.Security.Cryptography.Native/pal_err.cpp]: Do we expose these two ErrGetError and ErrGetErrorAlloc APIs because the only error code we need to know about is ERR_R_MALLOC_FAILURE, and this is a simpler approach than PAL'ing all error codes?

File: src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs [hunk 6]
```diff
@@ -371,7 +355,7 @@ private static void Disconnect(SafeSslHandle context)
             if (retVal < 0)
             {
                 //TODO (Issue #3362) check this error
-                libssl.SSL_get_error(context, retVal);
+                Crypto.SslGetError(context, retVal);
```
Comment (stephentoub) [inline src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs]: We're ignoring the result of this SslGetError call.  The TODO comment refers to issue 3362, which is closed.  3362 refers to issue 3954 as handling the rest, but 3954 appears to be specifically about channel binding.

**Comment by stephentoub** [comment]
LGTM

File: src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs [hunk 7]
```diff
@@ -9,29 +9,60 @@ internal static partial class Interop
...
-        [DllImport(Interop.Libraries.CryptoNative)]
-        internal static extern int SslRead(libssl.SafeSslHandle ssl, byte[] buf, int num);
+        [DllImport(Libraries.CryptoNative)]
+        internal static extern void SslDestroy(IntPtr ssl);
+
+        [DllImport(Libraries.CryptoNative)]
+        internal static extern void SslCtxDestroy(IntPtr ctx);
+
+        [DllImport(Libraries.CryptoNative)]
+        private static extern IntPtr SSLGetVersion(SafeSslHandle ssl);
```
Comment (bartonjs) [inline src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs]: Casing consistency: Everything else was "Ssl", this is "SSL"

File: src/Common/src/Microsoft/Win32/SafeHandles/SafeSslHandle.Unix.cs [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+                _writeBio.SetHandleAsInvalid(); // BIO got freed in SslDestroy
+            }
+            return true;
+        }
+
+        private SafeSslHandle() : base(IntPtr.Zero, true)
+        {
+        }
+    }
+}
```
Comment (bartonjs) [inline src/Common/src/Microsoft/Win32/SafeHandles/SafeSslHandle.Unix.cs]: Nit 1: No newline at EOF.
Nit 2: This class has a huge dependency on the functions in Interop.Ssl; I'd personally just put these SafeHandles in that same file.  Interop.Ssl requires this, this requires Interop.Ssl; sounds like a reasonable single-file construction to me.  (But still leaving it out here in neutral namespace-land, such as https://github.com/dotnet/corefx/blob/master/src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.ASN1.cs#L127)

File: src/Native/System.Security.Cryptography.Native/CMakeLists.txt [hunk 9]
```diff
@@ -26,6 +26,12 @@ if (OPENSSL_TLSV12)
     add_definitions(-DHAVE_TLS_V1_2=1)
 endif()
 
+include(CheckStructHasMember)
+check_struct_has_member ("SSL_CIPHER" algorithms openssl/ssl.h SSL_CIPHER_ALGORITHMS)
+if (SSL_CIPHER_ALGORITHMS)
+    add_definitions(-DHAVE_SSL_CIPHER_ALGORITHMS=1)
```
Comment (bartonjs) [inline src/Native/System.Security.Cryptography.Native/CMakeLists.txt]: I feel like other tests are checking for newer features, this one checks for an older one.

File: src/Native/System.Security.Cryptography.Native/pal_ssl.cpp (hunk 2 above)
Comment (bartonjs) [inline src/Native/System.Security.Cryptography.Native/pal_ssl.cpp]: I'm equally confused.  I see this for 0.9.8zg in ssl_locl.h.  If that's not automatically in the include chain, should we just include it?  At least for the 0.9.8 series?

File: src/Native/System.Security.Cryptography.Native/CMakeLists.txt (hunk 9 above)
Comment (eerhardt) [inline src/Native/System.Security.Cryptography.Native/CMakeLists.txt]: Agreed. I can switch it, but was having trouble with the naming. Any suggestions?
- HAVE_SSL_CIPHER_ALGORITHM_ENC
- HAVE_SSL_CIPHER_NO_ALGORITHMS
- HAVE_SSL_CIPHER_SPLIT_ALGORITHMS

File: src/Common/src/Microsoft/Win32/SafeHandles/SafeSslHandle.Unix.cs (hunk 8 above)
Comment (eerhardt) [inline src/Common/src/Microsoft/Win32/SafeHandles/SafeSslHandle.Unix.cs]: Will fix.

**Comment by bartonjs** [comment]
A few minor points, but otherwise LGTM.

File: src/Common/src/Interop/Unix/libssl/SslConnectionInfo.cs (hunk 4 above)
Comment (eerhardt) [inline src/Common/src/Interop/Unix/libssl/SslConnectionInfo.cs]: OK, I can change it back to int.

File: src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs (hunk 7 above)
Comment (eerhardt) [inline src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs]: Will fix.

File: src/Native/System.Security.Cryptography.Native/pal_ssl.cpp (hunk 2 above)
Comment (eerhardt) [inline src/Native/System.Security.Cryptography.Native/pal_ssl.cpp]: Yes, these values are in the source file ssl_locl.h. But neither that file nor those defines appear anywhere in the header files on my Mac.
Any suggestions?

File: src/Native/System.Security.Cryptography.Native/CMakeLists.txt (hunk 9 above)
Comment (bartonjs) [inline src/Native/System.Security.Cryptography.Native/CMakeLists.txt]: I'd probably go with "SPLIT", so it doesn't look weird at algorithm_mac

File: src/Native/System.Security.Cryptography.Native/pal_ssl.cpp (hunk 2 above)
Comment (bartonjs) [inline src/Native/System.Security.Cryptography.Native/pal_ssl.cpp]: > neither that file nor those defines appear anywhere in the header files on my Mac.

Ah, well, that's a good enough reason for me :)

File: src/Native/System.Security.Cryptography.Native/pal_err.cpp (hunk 5 above)
Comment (eerhardt) [inline src/Native/System.Security.Cryptography.Native/pal_err.cpp]: Yes, the MALLOC_FAILURE is the only error code we needed to know about. When we need to know about more, we can make a PAL enum for it.

File: src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs (hunk 6 above)
Comment (eerhardt) [inline src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs]: Opened #4031 for this TODO.

File: src/Native/System.Security.Cryptography.Native/pal_ssl.h [hunk 10]
```diff
@@ -6,6 +6,154 @@
...
+    SSL_eNULL = 32,
+    SSL_AES128 = 64,
+    SSL_AES256 = 128,
+    SSL_CAMELLIA128 = 256,
+    SSL_CAMELLIA256 = 512,
+    SSL_eGOST2814789CNT = 1024,
+    SSL_SEED = 2048,
+    SSL_AES128GCM = 4096,
+    SSL_AES256GCM = 8192
+#else
```
Comment (nguerrera) [inline src/Native/System.Security.Cryptography.Native/pal_ssl.h]: The declarations in the .h should be things that managed code can depend on unconditionally. There shouldn't be ifdefs changing enum values. If these are only used in the implementation and not expected to be known values to managed code, move them to the .cpp.

File: src/Native/System.Security.Cryptography.Native/pal_ssl.h [hunk 11]
```diff
@@ -6,6 +6,154 @@
...
+These values should be kept in sync with System.Security.Authentication.ExchangeAlgorithmType.
+*/
+enum class ExchangeAlgorithmType : int32_t
+{
+    None,
+    RsaSign = 9216,
+    RsaKeyX = 41984,
+    DiffieHellman = 43522,
+
+    //ExchangeAlgorithm constants which are not present in the managed ExchangeAlgorithmType enum.
```
Comment (nguerrera) [inline src/Native/System.Security.Cryptography.Native/pal_ssl.h]: Separate what you want managed code to match and what is implementation. Put implementation in .cpp.

File: src/Native/System.Security.Cryptography.Native/pal_ssl.h (hunk 10 above)
Comment (eerhardt) [inline src/Native/System.Security.Cryptography.Native/pal_ssl.h]: That makes sense. I'll move them to the .cpp as these values are only used by the native code, and not managed.

File: src/Native/System.Security.Cryptography.Native/pal_ssl.h (hunk 11 above)
Comment (eerhardt) [inline src/Native/System.Security.Cryptography.Native/pal_ssl.h]: This is an existing quirk in the code.  The public .NET enum has the first 4 values. But there are more values that OpenSSL can return than what is defined in the current public .NET enum. The existing managed code was already taking these int values and casting them to the enum.  See https://github.com/dotnet/corefx/blob/master/src/Common/src/Interop/Unix/libssl/SslConnectionInfo.cs#L138-L157

While I don't think this is a great approach, I didn't want to change the behavior with the shim work. I just ported the behavior to native.

@vijaykota - can you elaborate on the long term thoughts around these enums?

**Comment by eerhardt** [comment]
> I don't have problem switching to def/undef, but we should be consisten throughout.
This was easier to audit when we had everything in configure.cmake and pal_config.h.in with the same #cmakedefine01 mechanic.

I am actively working on this in a separate change (now that I am understanding CMake a little more...).  I plan on using #cmakedefin01, configure.cmake, config.h.in, like the existing files. These will be in separate, Crypto specific files.

**Comment by nguerrera** [comment]
> I am actively working on this in a separate change.

Thanks. :)

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by stephentoub** [comment]
@eerhardt, just FYI, this needs to be rebased to fix conflicts.

**Comment by eerhardt** [comment]
Rebased and responded to PR feedback.

File: src/Native/System.Security.Cryptography.Native/pal_ssl.cpp [hunk 12]
```diff
@@ -42,6 +44,343 @@ extern "C" SSL_CTX* SslCtxCreate(SSL_METHOD* method)
...
+    SSL_RC4 = 0x00020000L,
+    SSL_RC2 = 0x00040000L,
+    SSL_IDEA = 0x00080000L,
+    //SSL_eFZA = 0x00100000L,  this value is defined in ssl_locl.h, but has no match
+    SSL_eNULL = 0x00200000L,
+    SSL_AES = 0x04000000L,
+    SSL_CAMELLIA = 0x08000000L,
+    SSL_SEED = 0x10000000L,
+#endif
+};
```
Comment (nguerrera) [inline src/Native/System.Security.Cryptography.Native/pal_ssl.cpp]: General question: why do we have to hard-code these constants? Can't we get them from openssl headers? What guarantees/checks do we have to make sure they're right?

**Comment by nguerrera** [comment]
LGTM. I had one more question.

File: src/Native/System.Security.Cryptography.Native/pal_ssl.cpp (hunk 12 above)
Comment (eerhardt) [inline src/Native/System.Security.Cryptography.Native/pal_ssl.cpp]: See https://github.com/dotnet/corefx/pull/4026#discussion_r42600773

Honestly AFAIK, there isn't a way to check/guarantee these values since they aren't defined in headers on either my Linux or OSX machine. I sort of question whether what we are doing is a "public" API. The struct is visible, but it just exposes `unsigned long` members with no exposed defines for their values.

Comment (eerhardt) [inline src/Native/System.Security.Cryptography.Native/pal_ssl.cpp]: @vijaykota and @shrutigarg - do you have any ideas on this issue?

**Comment by stephentoub** [comment]
LGTM

File: src/Native/System.Security.Cryptography.Native/pal_ssl.cpp (hunk 12 above)
Comment (stephentoub) [inline src/Native/System.Security.Cryptography.Native/pal_ssl.cpp]: I'm going to go ahead and merge this.  If there's a better way to approach this, we can do it in a separate PR. Thanks!

Comment (nguerrera) [inline src/Native/System.Security.Cryptography.Native/pal_ssl.cpp]: I filed https://github.com/dotnet/corefx/issues/4074 to track that.

Comment (jasonwilliams200OK) [inline src/Native/System.Security.Cryptography.Native/pal_ssl.cpp]: Hey guys, maybe not exactly relevant to this discussion, but I upgraded a new mac as is to El Capitan (v10.11.1) and it wasn't succeeding `git clean -xdf; ./build.sh native` with latest AppleClang / Xcode (yesterday's updates) applied. I had to run `brew link --force openssl` (yeah i had to install brew first). Now it compiles 100% there as well. (more and moar platforms coverage is the spirit!) :)

Comment (eerhardt) [inline src/Native/System.Security.Cryptography.Native/pal_ssl.cpp]: Thanks, @jasonwilliams200OK .  See #3924 for more details on issues with El Capitan.

File: src/Native/System.Security.Cryptography.Native/pal_ssl.h (hunk 11 above)
Comment (vijaykota) [inline src/Native/System.Security.Cryptography.Native/pal_ssl.h]: Plan for future is to update the public contract to include new values. Tracked with https://github.com/dotnet/corefx/issues/3114

File: src/Native/System.Security.Cryptography.Native/pal_ssl.cpp (hunk 2 above)
Comment (vijaykota) [inline src/Native/System.Security.Cryptography.Native/pal_ssl.cpp]: Agree.. sounds reasonable to me. We had same problem with some other macros that are only available in the source code but not in public headers

File: src/Native/System.Security.Cryptography.Native/pal_ssl.cpp (hunk 12 above)
Comment (vijaykota) [inline src/Native/System.Security.Cryptography.Native/pal_ssl.cpp]: > since they aren't defined in headers on either my Linux or OSX machine

@eerhardt, we haven't found a way around this either
